Update translation guide (#5482)
ihhub/fheroes2: docs/TRANSLATION.md

`_n` function (a short name for `ngettext`) and `gettext_noop` function used only as indicator for the future string
translation. The string below, for example, can be translated in PO files:

```cpp
_( "Are you sure you want to quit?" )
```

## Adding new translations/localizations

If you want to add a new language localization, you will first need to add this language in the source code as a supported
language. Afterwards, a new PO file for that language will need to be added. It will have to be named according to the ISO
standard's two-character language abbreviations. To have font support for your language, you will have to specify what font
encoding/charset your language uses by adding it to the font generation code found in `/src/fheroes2/gui/ui_font.cpp`.
If a font supporting your language has not currently been implemented, then code for that will need to be added.

## Editing translations

We encourage you to use [**poedit**](https://poedit.net/) or [**gtranslator**](https://wiki.gnome.org/Apps/Gtranslator) to
edit translations.

Currently all implemented languages, except French, adhere to a standardized font encoding/charset.

> NB: The fheroes2 team has set a limit of 1000 lines of added lines for any PR for translations. This is because every PR needs

Author: Nota bene, translates to "Note well". I can change it to "Note:" to make it more obvious.

to be reviewed and adding too many lines at once will only slow this process down. In addition, Github becomes hard to navigate
once too many changes, comments etc. are present within a PR page, further slowing down the process of reviewing it.

> Preferrably a PR should contain a small amount of changes, like 100, all focused on translating a few parts of the game like

Reviewer: `like 100` --> `about 100 lines`.

> Preferrably a PR should contain a small amount of changes, like 100, all focused on translating a few parts of the game like

Reviewer: `few parts of the game like` --> `few parts of the game, for example,`

creature names or castle buildings.

## Build binary translation files

Once the translation files have been modified, for Linux/MacOS run the `make` command below in `/files/lang` to create
machine object (MO) binary files which can be used by the fheroes2 engine.

> For the German de.po, this would be the command:

Reviewer: For example, for the German de.po, this would be the command:

> 

Reviewer: Please remove this empty line.

```bash
make de.mo
```

This MO file should then be put in the `/files/lang` folder used by the fheroes2 executable, in other words not the one located in

> the source directory.

Reviewer: This sentence is confusing. Isn't for Linux / MacOS developers / translator the game is located in the same directory as the source code?

Author: I was very unsure about what to write here because I don't know what the folder structure is for the various systems. I only know that for Windows the MO files go into the `/files/lang` folder in the same directory as the `fheroes2.exe` executable.

Author: The point here is that someone could be confused about why their compiled MO file isn't showing up in the game when the problem is that they have put it in the wrong `/files/lang/`.

Reviewer 2: @zenseii Not really yet. You see, currently fheroes2 is available on macOS via 3rd-party package managers, such as MacPorts or Homebrew, and every of them have its own layout. May be in future, when we will be able to distribute fheroes2 as Mac app bundle (we are currently able to build it, but not able to sign & notarize), app will always be installed to `/Applications`, how it is accustomed to do on macOS.

> For Windows users that use POEdit or similar, they can compile with that program. However, note that the program will need to be set

Reviewer: `POEdit or similar application`

to compile the MO file in the font encoding/Charset that the language that you are translating to has been set to.

For example, for German you will have to set font encoding to CP1252, while for Russian this would be CP1251. Later when submitting
a PR with your changes, you will have to save the PO file in UTF-8 encoding because this is what Github supports.

## Updating PO templates and translatable strings in PO files

Currently all PO files are automatically updated with new strings after each commit that brings changes to the ingame text.
Should you still need to update strings locally, this can be acheived by running the command below in `/src/dist` to generate
a new portable object template (POT) file. Windows users will need to setup an environment that lets them run `make`, like

> Windows Subsystem for Linux (WSL) or [Cygwin](https://www.cygwin.com/).

Reviewer 3: ```suggestion
Windows Subsystem for Linux (WSL) or [Cygwin](https://www.cygwin.com/)/[MSYS2](https://www.msys2.org/).
```

This is minor, but still...

Author: Sure, why not! I imagine there is no reason this wouldn't work with msys2? I personally only have experience using gettext with Cygwin.

`_n` function (a short name for `ngettext`) and `gettext_noop` function used only as indicator for the future string
translation. The string below, for example, can be translated in PO files:

```cpp
_( "Are you sure you want to quit?" )
```

## Adding new translations/localizations

If you want to add a new localization, you will first need to add it to the `SupportedLanguage` list/enumeration in the source code.
Afterwards, a new PO file for it will need to be added. It will have to be named according to the ISO standard's two-character
language abbreviations. Then to have font support, you will have to specify what font encoding/charset is compatible by adding
the language to the font generation code found in `/src/fheroes2/gui/ui_font.cpp`. If a compatible font encoding has not currently
been implemented, then code for that will need to be written.

## Editing translations

We encourage you to use [**poedit**](https://poedit.net/) or [**gtranslator**](https://wiki.gnome.org/Apps/Gtranslator) to
edit translations.

Currently all implemented languages, except French, adhere to a standardized font encoding/charset.

> NOTE: The fheroes2 team has set a limit of 1000 added or modified lines for any Pull Request(PR) for translations. This is because

Reviewer: Please add a space before (PR).

Author: ```suggestion
NOTE: The fheroes2 team has set a limit of 1000 added or modified lines for any Pull Request (PR) for translations. This is because

```

to navigate once too many changes, comments and so on are present within the same PR page, further slowing down the process of reviewing it.

Preferrably a PR should contain a small amount of changes, about 100 lines, all focused on translating a few parts of the game - for
example creature names or castle buildings.

## Build binary translation files

Once the translation files have been modified, for Linux/MacOS run the `make` command below in `/files/lang` to create
machine object (MO) binary files which can be used by the fheroes2 engine.

For exmaple, for the German PO file, `de.po`, the following would be the command:
```bash
make de.mo
```

To have this MO file used by the engine, it should then be placed in the `/files/lang` folder used by the fheroes2 executable.
The exact location of this folder depends on the operating system. On Windows, it is usually located in the app installation
directory. On Linux, it is usually located in the `/usr/share/fheroes2` or `/usr/local/share/fheroes2`. Currently for MacOS
users this location is dependent on what third-party package manager is used to install fheroes2.

> For Windows users that use POEdit or a similar application, they can compile with that program. However, note that the program will

Reviewer: that --> who.

Author: ```suggestion
For Windows users who use POEdit or a similar application, they can compile with that program. However, note that the program will

```

need to be set to compile the MO file in the font encoding/Charset that the language that you are translating to has been set to.

For example, for German you will have to set font encoding to CP1252, while for Russian this would be CP1251. Later when submitting
a PR with your changes, you will have to save the PO file in UTF-8 encoding because this is what Github supports.

## Updating PO templates and translatable strings in PO files

Currently all PO files are automatically updated with new strings after each commit that brings changes to the ingame text.

> Should you still need to update strings locally, this can be acheived by running the command below in `/src/dist` to generate

Reviewer: Acheived --> achieved.

> Should you still need to update strings locally, this can be acheived by running the command below in `/src/dist` to generate

Reviewer: Please replace first should by if to make the sentence as a statement.

Reviewer: This is tricky. Should might work as well🤔

Author: ```suggestion
If for whatever reason you still need to update strings locally, this can be acheived by running the command below in `/src/dist` to generate

```

Author: Hi, @ihhub!
Maybe the suggestion above sounds better "for whatever reason" can be removed too or "you" can be placed before that.

# [**fheroes2**](README.md) translation guide

> This project uses portable object (PO) files to handle localization in various languages. The PO files are located in `/files/lang`.

Reviewer 2: This looks like they are located in the `[filesystem root]/files/lang` :)

```suggestion
This project uses portable object (PO) files to handle localization in various languages. The PO files are located in the `files/lang`
subdirectory of the project source tree.
```

The current instruction is designed for Linux, MacOS and Windows users.

## Finding translatable strings in the codebase

Translatable strings can be found in the source code as arguments to the `_` function (a short name for `gettext`),
`_n` function (a short name for `ngettext`) and `gettext_noop` function used only as indicator for the future string
translation. The string below, for example, can be translated in PO files:

```cpp
_( "Are you sure you want to quit?" )
```

## Adding new translations/localizations

If you want to add a new localization, you will first need to add it to the `SupportedLanguage` list/enumeration in the source code.
Afterwards, a new PO file for it will need to be added. It will have to be named according to the ISO standard's two-character
language abbreviations. Then to have font support, you will have to specify what font encoding/charset is compatible by adding

> the language to the font generation code found in `/src/fheroes2/gui/ui_font.cpp`. If a compatible font encoding has not currently

Reviewer 2: ```suggestion
the language to the font generation code found in `src/fheroes2/gui/ui_font.cpp`. If a compatible font encoding has not currently
```

been implemented, then code for that will need to be written.

## Editing translations

We encourage you to use [**poedit**](https://poedit.net/) or [**gtranslator**](https://wiki.gnome.org/Apps/Gtranslator) to
edit translations.

Currently all implemented languages, except French, adhere to a standardized font encoding/charset.

NOTE: The fheroes2 team has set a limit of 1000 added or modified lines for any Pull Request (PR) for translations. This is because

every PR needs to be reviewed and adding too many lines at once will only slow this process down. In addition, GitHub becomes hard
to navigate once too many changes, comments and so on are present within the same PR page, further slowing down the process of reviewing it.

Preferrably a PR should contain a small amount of changes, about 100 lines, all focused on translating a few parts of the game - for
example creature names or castle buildings.

## Build binary translation files

> Once the translation files have been modified, for Linux/MacOS run the `make` command below in `/files/lang` to create

Reviewer 2: ```suggestion
Once the translation files have been modified, for Linux/MacOS run the `make` command below in the `files/lang` subdirectory to create
```

machine object (MO) binary files which can be used by the fheroes2 engine.

For exmaple, for the German PO file, `de.po`, the following would be the command:
```bash
make de.mo
```

> To have this MO file used by the engine, it should then be placed in the `/files/lang` folder used by the fheroes2 executable.

Reviewer 2: ```suggestion
To have this MO file used by the engine, it should then be placed in the `files/lang` folder used by the fheroes2 executable.
```

The exact location of this folder depends on the operating system. On Windows, it is usually located in the app installation
directory. On Linux, it is usually located in the `/usr/share/fheroes2` or `/usr/local/share/fheroes2`. Currently for MacOS
users this location is dependent on what third-party package manager is used to install fheroes2.

For Windows users who use POEdit or a similar application, it is possible to compile the MO file using such a program. However, note that the program will

need to be set to compile the MO file in the font encoding/Charset that the language that you are translating to has been set to.

For example, for German you will have to set font encoding to CP1252, while for Russian this would be CP1251. Later when submitting
a PR with your changes, you will have to save the PO file in UTF-8 encoding because this is what Github supports.

## Updating PO templates and translatable strings in PO files

Currently all PO files are automatically updated with new strings after each commit that brings changes to the ingame text.

> If for whatever reason you still need to update strings locally, this can be achieved by running the command below in `/src/dist` to generate

Reviewer 2: ```suggestion
If for whatever reason you still need to update strings locally, this can be achieved by running the command below in `src/dist` to generate
```

a new portable object template (POT) file. Windows users will need to setup an environment that lets them run `make`, like
Windows Subsystem for Linux (WSL) or [Cygwin](https://www.cygwin.com/)/[MSYS2](https://www.msys2.org/).

```bash
make pot
```

> Once the POT file has been created, go to `/files/lang` and run the command below to update translatable strings in the PO files.

Reviewer 2: ```suggestion
Once the POT file has been created, go to the `files/lang` folder and run the command below to update translatable strings in the PO files.
```

Preferrably a PR should contain a small amount of changes, like 100, all focused on translating a few parts of the game like creature names or castle buildings.

## Build binary translation files

Once the translation files have been modified, for Linux/MacOS run the `make` command below in `/files/lang` to create machine object (MO) binary files which can be used by the fheroes2 engine.

For the German de.po, this would be the command:

```bash
make de.mo
```

This MO file should then be put in the `/files/lang` folder used by the fheroes2 executable, in other words not the one located in the source directory.

For Windows users that use POEdit or similar, they can compile with that program. However, note that the program will need to be set to compile the MO file in the font encoding/Charset that the language that you are translating to has been set to.

For example, for German you will have to set font encoding to CP1252, while for Russian this would be CP1251. Later when submitting a PR with your changes, you will have to save the PO file in UTF-8 encoding because this is what Github supports.

## Updating PO templates and translatable strings in PO files

> Currently all PO files are automatically updated with new strings after each commit that brings changes to the ingame text. Should you still need to update strings locally, this can be acheived by running the command below in `/src/dist` to generate a new portable object template (POT) file. Windows users will need to setup an environment that lets them run 'make', like Windows Subsystem for Linux (WSL) or [Cygwin](https://www.cygwin.com/).

Reviewer 2: Please replace 'make' by \`make\` to match the representation of other commands.

```bash
make pot
```

Once the POT file has been created, go to `/files/lang` and run the command below to update translatable strings in the PO files. If you are using programs mentioned above like POEdit, then they have options to merge new strings from a POT file.

```bash
make merge
```

> Alternatively, to update a specific language you can run the following command, using the French PO file `fr.po` as an example:

Author: I see! I must stay more up to date then. Thank you for explaining this.
